### ISSUE TITLE
How to attach hook to constructor?

### ISSUE THREAD

**Issue by davout1806** [mjancarik/to-aop#2]
How, if possible, to attach a hook to a class's constructor?

I've tried 'constructor' and the name of the class with no apparent effect. Thanks

```
const constructorBefore = createHook(
  hookName.beforeMethod,
  'constructor',
  ({ target, object, property, context, args, meta }) => {
    console.log(args)
  }
)
```

```
const constructorBefore = createHook(
  hookName.beforeMethod,
  'EffectiveConfigFiles',
  ({ target, object, property, context, args, meta }) => {
    console.log(args)
  }
)
```

**Comment by mjancarik** [comment]
Hi,

It's not possible now. I disabled that because I had problem with consistency behaviour. But I rewrote some parts for 0.5.0 version. I tried quick fix and hooks work well for constructor now. Could you send me all your code snippet with calling `aop` method? I will try it in tests and than I will release new version.

Thanks.

**Comment by davout1806** [comment]
I'm using your code to apply design by contract checks and (soon) logging. So I can always do those directly inside the constructors themselves. I just prefer not to clutter my classes with that stuff - hence AOP.

Thanks.

```
"use strict";

const { default: ow } = require('ow')
const { aop, hookName, createHook } = require('to-aop')
const EffectiveConfigSelector = require('../EffectiveConfigSelector')

const constructorBefore = createHook(
  hookName.beforeMethod,
  'EffectiveConfigSelector',
  ({ target, object, property, context, args, meta }) => {
    ow(args[0], 'settingsObj', ow.object.partialShape({
      root: ow.string,
      mod_file_folder: ow.string
    }))
  }
)

aop(EffectiveConfigSelector, constructorBefore)
```
```
class EffectiveConfigSelector {

  constructor(settings) {
    this.settingsObj = settings
  }
```

**Comment by mjancarik** [comment]
Hi,

I have prepared version for you but It has one side effect that you must calling new for returned class from `aop` method. Is it a problem?

```
"use strict";

const { default: ow } = require('ow')
const { aop, hookName, createHook } = require('to-aop')
const EffectiveConfigSelector = require('../EffectiveConfigSelector')

const constructorBefore = createHook(
  hookName.beforeMethod,
  'constructor',
  ({ target, object, property, context, args, meta }) => {
    ow(args[0], 'settingsObj', ow.object.partialShape({
      root: ow.string,
      mod_file_folder: ow.string
    }))
  }
)

let HookedEffectiveConfigSelector = aop(EffectiveConfigSelector, constructorBefore, { constructor: true })

new HookedEffectiveConfigSelector({ some: 'settings' })
```

**Comment by davout1806** [comment]
So the only change would be
```
let HookedEffectiveConfigSelector = aop(EffectiveConfigSelector, constructorBefore, { constructor: true })

new HookedEffectiveConfigSelector({ some: 'settings' })
```

I'd be fine with that. For non-constructors would I continue with the current way or would I do something similar to above? 
```
const _basePathsModsSpecificallyExcludeAfter = createHook(
  hookName.afterMethod, 
  '_basePathsModsSpecificallyExclude',
  ({ target, object, property, context, args, payload, meta }) => {
    ow(payload, 'return value', ow.array)
  })

//This
aop(EffectiveConfigSelector, _basePathsModsSpecificallyExcludeAfter)

// or something like 
let HookedEffectiveConfigSelector = aop(EffectiveConfigSelector, _basePathsModsSpecificallyExcludeAfter, { constructor: false })

new HookedEffectiveConfigSelector({ some: 'settings' })
```

BTW thanks for working on this.

**Comment by mjancarik** [comment]
Hi,

I released new version to-aop@0.5.2.

Yes, you must only change code for constructors hook (snippet 1). For non-constructors you should continue with the current way. 

Please let me know if everything works.

**Comment by davout1806** [comment]
Made the change. Everything seems to work!

Thanks
